crates/cadenza-eval/src/interner.rs

```diff
@@ -1,141 +1,463 @@
+            map: rustc_hash::FxHashMap::default(),
+            strings: Vec::new(),
+        }
     }
 
-    /// Interns a string, returning its unique ID.
-    ///
-    /// If the string has already been interned, returns the existing ID.
-    /// Otherwise, assigns a new ID and stores the string.
-    pub fn intern(&mut self, s: &str) -> InternedId {
-        if let Some(&id) = self.map.get(s) {
-            return id;
+    fn insert(&mut self, s: &str) -> u32 {
+        if let Some(&index) = self.map.get(s) {
+            return index;
         }
-
-        let id = InternedId(self.strings.len() as u32);
+        let index = self.strings.len() as u32;
         let owned = s.to_string();
         self.strings.push(owned.clone());
-        self.map.insert(owned, id);
-        id
+        self.map.insert(owned, index);
+        index
     }
 
-    /// Looks up the string for an interned ID.
-    ///
-    /// # Panics
-    ///
-    /// Panics if the ID was not created by this interner.
-    pub fn resolve(&self, id: InternedId) -> &str {
-        &self.strings[id.0 as usize]
+    fn get(&self, index: u32) -> &str {
+        &self.strings[index as usize]
+    }
+}
+
+static STRING_STORAGE: OnceLock<std::sync::Mutex<StringData>> = OnceLock::new();
+
+fn string_storage() -> &'static std::sync::Mutex<StringData> {
+    STRING_STORAGE.get_or_init(|| std::sync::Mutex::new(StringData::new()))
+}
+
+impl Storage for Strings {
+    type Index = u32;
+    type Value = str;
+
+    fn insert(value: &str) -> Self::Index {
+        string_storage().lock().unwrap().insert(value)
     }
 
-    /// Tries to look up an existing interned ID without creating a new one.
-    pub fn get(&self, s: &str) -> Option<InternedId> {
-        self.map.get(s).copied()
+    fn resolve(index: Self::Index) -> &'static Self::Value {
+        // SAFETY: Strings are never removed from storage, so the reference is valid
+        // for the lifetime of the program.
+        let storage = string_storage().lock().unwrap();
+        let s = storage.get(index);
+        unsafe { std::mem::transmute::<&str, &'static str>(s) }
     }
+}
+
+/// Type alias for interned strings.
+pub type InternedString = Interned<Strings>;
+
+// =============================================================================
+// Integer Storage
+// =============================================================================
 
-    /// Returns the number of interned strings.
-    pub fn len(&self) -> usize {
-        self.strings.len()
+/// Storage for interned integer literals.
+///
+/// Integer literals are stored as `Result<i64, ()>` where `Err(())` indicates
+/// a parse error. This allows us to intern the literal string and cache the
+/// parse result.
+#[derive(Debug, Clone, Copy, Default)]
+pub struct Integers {
+    _not_send_sync: PhantomData<*const ()>,
+}
+
+/// Internal storage data for integers.
+struct IntegerData {
+    map: rustc_hash::FxHashMap<String, u32>,
+    values: Vec<Result<i64, ()>>,
```

Comment:
We should store the actual error so we at least get a message when resolving the value: https://doc.rust-lang.org/std/num/struct.ParseIntError.html.